horovod/spark/common/util.py

```diff
@@ -0,0 +1,546 @@
+    elif isinstance(validation, str):
+        bool_dtype = metadata[validation]['spark_data_type'] == BooleanType
+        val_df = train_df.filter(
+            f.col(validation) if bool_dtype else f.col(validation) > 0).drop(validation)
+        train_df = train_df.filter(
+            ~f.col(validation) if bool_dtype else f.col(validation) == 0).drop(validation)
+
+        # Approximate ratio of validation data to training data for proportionate scale
+        # of partitions
+        timeout_ms = 1000
+        confidence = 0.90
+        train_rows = train_df.rdd.countApprox(timeout=timeout_ms, confidence=confidence)
+        val_rows = val_df.rdd.countApprox(timeout=timeout_ms, confidence=confidence)
+        validation_ratio = val_rows / (val_rows + train_rows)
+    elif validation:
+        raise ValueError('Unrecognized validation type: {}'.format(type(validation)))
+
+    return train_df, val_df, validation_ratio
+
+
+@contextlib.contextmanager
+def prepare_data(num_processes, store, df, label_columns, feature_columns,
+                 validation=None, sample_weight_col=None,
+                 partitions_per_process=10, verbose=False):
+    if num_processes <= 0 or partitions_per_process <= 0:
+        raise ValueError('num_proc={} and partitions_per_process={} must both be > 0'
+                         .format(num_processes, partitions_per_process))
+
+    if not label_columns:
+        raise ValueError('Parameter label_columns cannot be None or empty')
+
+    num_partitions = num_processes * partitions_per_process
+    if verbose:
+        print('num_partitions={}'.format(num_partitions))
+
+    for col in label_columns:
+        if col not in df.columns:
+            raise ValueError('Label column {} does not exist in this DataFrame'.format(col))
+
+    if feature_columns is None:
+        feature_columns = [col for col in df.columns if col not in set(label_columns)]
+
+    key = _training_cache.create_key(df, store, validation)
+    with _training_cache.lock:
+        _training_cache.set_in_use(key, True)
+        if _training_cache.is_cached(key, store):
+            dataset_idx = _training_cache.get_dataset(key)
+            train_rows, val_rows, metadata, avg_row_size = _training_cache.get_dataset_properties(dataset_idx)
+            train_data_path = store.get_train_data_path(dataset_idx)
+            val_data_path = store.get_val_data_path(dataset_idx)
+            if verbose:
+                print('using cached dataframes for key: {}'.format(key))
+                print('train_data_path={}'.format(train_data_path))
+                print('train_rows={}'.format(train_rows))
+                print('val_data_path={}'.format(val_data_path))
+                print('val_rows={}'.format(val_rows))
+        else:
+            dataset_idx = _training_cache.next_dataset_index(key)
+            train_data_path = store.get_train_data_path(dataset_idx)
+            val_data_path = store.get_val_data_path(dataset_idx)
```

Comment:
The `train_data_path ` and `val_data_path ` should be appended by `filesystem_prefix`, by calling `store.get_full_path` ?

Reply:
Yes, good point.  Fixed.  See above for details.